src/facade/reply_builder.cc

```diff
@@ -228,34 +249,55 @@ void SinkReplyBuilder2::WritePiece(std::string_view str) {
 
 void SinkReplyBuilder2::WriteRef(std::string_view str) {
   NextVec(str);
+  ref_indices_.push_back(vecs_.size() - 1);
   total_size_ += str.size();
 }
 
 void SinkReplyBuilder2::Flush() {
-  auto ec = sink_->Write(vecs_.data(), vecs_.size());
-  if (ec)
+  auto& reply_stats = tl_facade_stats->reply_stats;
+
+  send_active_ = true;
+  uint64_t before_ns = util::fb2::ProactorBase::GetMonotonicTimeNs();
+  reply_stats.io_write_cnt++;
+  reply_stats.io_write_bytes += total_size_;
+
+  if (auto ec = sink_->Write(vecs_.data(), vecs_.size()); ec)
     ec_ = ec;
 
+  uint64_t after_ns = util::fb2::ProactorBase::GetMonotonicTimeNs();
+  reply_stats.send_stats.count++;
+  reply_stats.send_stats.total_duration += (after_ns - before_ns) / 1'000;
+  send_active_ = false;
+
+  if (buffer_.InputLen() * 2 > buffer_.Capacity())  // If needed, grow backing buffer
+    buffer_.Reserve(std::min(kMaxBufferSize, buffer_.Capacity() * 2));
+
+  total_size_ = 0;
   buffer_.Clear();
   vecs_.clear();
-  total_size_ = 0;
+  ref_indices_.clear();
 }
 
 void SinkReplyBuilder2::FinishScope() {
-  // If batching or aggregations are not enabled, flush
-  Flush();
-
-  // TODO: otherwise iterate over vec_ and copy items to buffer_
-  // whilst also updating their pointers
-}
-
-bool SinkReplyBuilder2::IsInBuf(const void* ptr) const {
-  auto ib = buffer_.InputBuffer();
-  return ptr >= ib.data() && ptr <= ib.data() + ib.size();
+  if (!batched_ || total_size_ * 2 >= kMaxBufferSize)
+    return Flush();
+
+  // Check if we have enough space to copy all refs to buffer
+  size_t ref_bytes = total_size_ - buffer_.InputLen();
+  if (ref_bytes > buffer_.AppendLen())
+    return Flush();
+
+  // Copy all extenral references to buffer to safely keep batching
+  for (unsigned i : ref_indices_) {
```

Comment:
I understand now the motivation but I still do not understand why ref_indices_ are needed. From the code it seems they contain `{0, 1, ..n}` sequence. I am probably wrong, but not sure why :)

Reply:
ah, I understand. they are cleared here, so the next sequence will start from K. still seems that they represent a range `[i, j]` so you do not really need a vector for that.

Reply:
No, suppose the following case. You write lots of small refs and decide to batch them once the scope ends, so they turn into pieces. Now you write some other array and add a ref at the very end of vec_. To find it, you need to traverse the whole array

Reply:
They represent some kind of range with at least 50% of the range being external refs, I can simplify that

Reply:
Note that I only add the index for refs, not for pieces, but two pieces are never adjacent (except for scope end batching)